### PR TITLE
Registration of rules and getter for fetching the rules

### PR DESCRIPTION
**Pull request by gulshanvasnani**
PR contains following :-
1. Method to register rules(`registerRule`).
2. Method to fetch rule by name and rule address(`getRuleByName` and `getRuleByAddress`).
3. Exposed TokenRules helper through index.js.
4. Added license information.

Fixes:- #17 #18

### CODE REVIEW

File: lib/helper/TokenRules.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+   * @return Promise object.
+   */
+  registerRule(ruleName, ruleAddress, ruleAbi, txOptions) {
+    const oThis = this;
+
+    const txObject = oThis._registerRule(ruleName, ruleAddress, ruleAbi, txOptions);
+
+    let txReceipt;
+
+    return txObject
```
Comment (abhayks1) [inline lib/helper/TokenRules.js]: We should integrate this with TxSender.

File: lib/helper/TokenRules.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  /**
+   * Private method which is used to register a custom rule to the economy.
+   *
+   * @param ruleName Name of the rule.
+   * @param ruleAddress Contract address of the rule.
+   * @param ruleAbi Abi of the rule.
+   * @param txOptions Tx options.
+   * @private
+   */
+  _registerRule(ruleName, ruleAddress, ruleAbi, txOptions) {
```
Comment (abhayks1) [inline lib/helper/TokenRules.js]: Rename it to _registerRuleRawTx

File: lib/helper/TokenRules.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    const contract = new oThis.auxiliaryWeb3.eth.Contract(jsonInterface, oThis.tokenRules, txOptions);
+    return contract.methods.registerRule(ruleName, ruleAddress, ruleAbi);
+  }
+
+  /**
+   * It is used to fetch rule data by its name.
+   *
+   * @param ruleName Name of the rule.
+   * @param txOptions Tx options.
+   * @returns Rule data if present.
```
Comment (abhayks1) [inline lib/helper/TokenRules.js]: Let's document what are the datapoints in Rule object.

File: lib/helper/TokenRules.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+    return rule;
+  }
+
+  /**
+   * It is used to fetch rule data by its address.
+   *
+   * @param ruleAddress Address of the rule contract.
+   * @param txOptions Tx options.
+   * @returns Rule data if present.
```
Comment (abhayks1) [inline lib/helper/TokenRules.js]: Let's document what are the datapoints in Rule object.

File: lib/helper/TokenRules.js [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+   */
+  async getRuleByAddress(ruleAddress, txOptions) {
+    const oThis = this;
+
+    const jsonInterface = oThis.abiBinProvider.getABI(contractName),
+      contract = new oThis.auxiliaryWeb3.eth.Contract(jsonInterface, oThis.tokenRules, txOptions);
+
+    const ruleIndex = await contract.methods.rulesByAddress(ruleAddress).call(),
+      rule = await contract.methods.rules(ruleIndex.index).call();
```
Comment (abhayks1) [inline lib/helper/TokenRules.js]: rule => "const rule"

File: lib/helper/TokenRules.js [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    const oThis = this;
+
+    const jsonInterface = oThis.abiBinProvider.getABI(contractName),
+      contract = new oThis.auxiliaryWeb3.eth.Contract(jsonInterface, oThis.tokenRules, txOptions);
+
+    const ruleNameHash = oThis.auxiliaryWeb3.utils.soliditySha3({ t: 'string', v: ruleName }),
+      ruleIndex = await contract.methods.rulesByNameHash(ruleNameHash).call(),
+      rule = await contract.methods.rules(ruleIndex.index).call();
+
+    return rule;
```
Comment (abhayks1) [inline lib/helper/TokenRules.js]: rule => "const rule"

File: lib/helper/TokenRules.js (hunk 3 above)
Comment (abhayks1) [reply lib/helper/TokenRules.js]: Rule object: {
        ruleName;
        ruleAddress;
        ruleAbi;
    }

File: test/integration/ExecuteRule.js [hunk 7]
```diff
@@ -153,4 +155,36 @@ describe('ExecuteRule', async function() {
...
+    assert.strictEqual(response.events.RuleRegistered['returnValues']._ruleAddress, mockRuleAddress);
+
+    // Verify the rule data using rule name.
+    const ruleByNameData = await rules.getRuleByName(mockRule, txOptions);
+    assert.strictEqual(ruleByNameData.ruleName, mockRule, 'Incorrect rule name was registered');
+    assert.strictEqual(ruleByNameData.ruleAddress, mockRuleAddress, mockRuleAddress, 'Incorrect rule address');
+
+    // Verify the rule data using rule address.
+    const ruleByAddressData = await rules.getRuleByAddress(mockRuleAddress, txOptions);
+    assert.strictEqual(ruleByNameData.ruleName, mockRule, 'Incorrect rule name was registered');
```
Comment (abhayks1) [inline test/integration/ExecuteRule.js]: ruleByNameData => ruleByAddressData

File: test/integration/ExecuteRule.js [hunk 8]
```diff
@@ -153,4 +155,36 @@ describe('ExecuteRule', async function() {
...
+    assert.strictEqual(ruleByNameData.ruleAddress, mockRuleAddress, mockRuleAddress, 'Incorrect rule address');
```
Comment (abhayks1) [inline test/integration/ExecuteRule.js]: ruleByNameData => ruleByAddressData

File: lib/helper/TokenRules.js (hunk 2 above)
Comment (abhayks1) [inline lib/helper/TokenRules.js]: Rename _registerRule to _registerRuleRawTx

File: lib/helper/TokenRules.js (hunk 3 above)
Comment (gulshanvasnani) [reply lib/helper/TokenRules.js]: Good point 👍

**Pull request merged**